sklearn/utils/extmath.py

```diff
@@ -264,6 +265,55 @@ def randomized_range_finder(A, size, n_iter=2,
     Q, _ = linalg.qr(safe_sparse_dot(A, Q),  mode='economic')
     return Q
 
+def svd_solver(M, n_components, svd_method='auto', n_iter=2, random_state=0):
+    """
+    TODO: Documentation
+    """
+
+    legal_svd_methods = ['auto', 'full', 'arpack', 'randomized']
+
+    if svd_method not in legal_svd_methods:
+        raise ValueError("Unknown SVD method: '{0}'. svd_method must be one of"
+                         " {1}.".format(svd_method, legal_svd_methods))
+
+    random_state = check_random_state(random_state)
+    n_samples, n_features = M.shape
+
+    if svd_method == 'auto':
+        # Small problem, just call full SVD
+        if max(M.shape) <= 500:
+            svd_method = 'full'
+        elif n_components >= 1 and n_components < .8 * min(M.shape):
+            svd_method = 'randomized'
+        # This is also the case of n_components in (0,1)
+        else:
+            svd_method = 'full'
+
+    if svd_method == 'full':
+        # Center data
+        mean = np.mean(M, axis=0)
+        M -= mean
```

Comment:
That would do the job, but we must be also be able to avoid the copy when it is unnecessary. Namely, if the caller already does the copy, the SVD solver should be called with `copy=False`.